texar/torch/data/tokenizers/sentencepiece_tokenizer.py

```diff
@@ -0,0 +1,256 @@
+              cache_dir: Optional[str] = None) -> str:
+        if cache_dir is None:
+            cache_path = str(default_download_dir('SentencePiece'))
+        else:
+            if not os.path.isdir(cache_dir):
+                raise ValueError(f"Cache directory ({cache_dir}) should be a "
+                                 f"directory.")
+            cache_path = os.path.abspath(cache_dir)
+
+        maybe_create_dir(cache_path)
+
+        spm.SentencePieceTrainer.Train(cmd)
+        cwd = os.getcwd()
+
+        vocab_file = os.path.join(cwd, cls._VOCAB_FILE_NAMES['vocab_file'])
+        out_vocab_file = os.path.join(
+            cache_path, cls._VOCAB_FILE_NAMES['vocab_file'])
+
+        if os.path.abspath(vocab_file) != os.path.abspath(out_vocab_file):
+            move(vocab_file, out_vocab_file)
+
+        # Delete spiece.vocab (We might want to keep it as well)
+        extra_file = vocab_file.rstrip('model') + 'vocab'
+        os.remove(extra_file)
+
+        return cache_path
+
+    # spm.SentencePieceProcessor() is a SwigPyObject object which cannot be
+    # pickled. We need to define __getstate__ here.
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state["sp_model"] = None
+        state["vocab_file"] = None
+        return state, self.vocab_file
+
+    # spm.SentencePieceProcessor() is a SwigPyObject object which cannot be
+    # pickled. We need to define __setstate__ here.
+    def __setstate__(self, d):
+        self.__dict__, self.vocab_file = d
+        self.sp_model = spm.SentencePieceProcessor()
+        self.sp_model.Load(self.vocab_file)
+
+    def save_vocab(self, save_directory: str) -> Tuple[str]:
+        r"""Save the sentencepiece vocabulary (copy original file) to
+        a directory.
+        """
+        if not os.path.isdir(save_directory):
+            raise ValueError("Vocabulary path ({}) should be a "
+                             "directory".format(save_directory))
+        out_vocab_file = os.path.join(
+            save_directory, self._VOCAB_FILE_NAMES['vocab_file'])
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
+
+        return (out_vocab_file,)
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.sp_model)
+
+    def _map_text_to_token(self, text: str) -> List[str]:  # type: ignore
+        return self.sp_model.EncodeAsPieces(text)
+
+    def _map_token_to_id(self, token: str) -> int:
+        return self.sp_model.PieceToId(token)
+
+    def _map_id_to_token(self, index: int) -> str:
+        token = self.sp_model.IdToPiece(index)
+        return token
+
+    def map_token_to_text(self, tokens: List[str]) -> str:
+        r"""Maps a sequence of tokens (string) in a single string."""
+        out_string = self.sp_model.DecodePieces(tokens)
+        return out_string
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
```

Comment:
What's the behavior of special tokens in this tokenizer? Will they be added to the input vocab? Or does the input vocab must already include those tokens? Make it clear in the docstring